File: StrukturDataStack/stack_expression_linkedlist.py
import re


class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class StackLL:
    def __init__(self):
        self.top = None

    def is_empty(self):
        return self.top is None

    def push(self, data):
        new = Node(data)
        new.next = self.top
        self.top = new

    def pop(self):
        if self.is_empty():
            raise Exception("Stack kosong!")
        value = self.top.data
        self.top = self.top.next
        return value

    def peek(self):
        if self.is_empty():
            return None
        return self.top.data

    def print_stack(self):
        temp = self.top
        print("[ ", end="")
        while temp:
            print(temp.data, end=" ")
            temp = temp.next
        print("]")


# ================= OPERATOR =================
def precedence(op):
    if op == "^": return 3
    if op in ["*", "/"]: return 2
    if op in ["+", "-"]: return 1
    return 0


def is_operator(op):
    return op in ["+", "-", "*", "/", "^"]


# ================= TOKENIZER =================
def tokenize(expr):
    tokens = re.findall(r'\d+\.?\d*|[()+\-*/^]', expr)

    # handle unary minus
    result = []
    i = 0
    while i < len(tokens):
        if tokens[i] == "-" and (i == 0 or tokens[i-1] in "(+*/^-"):
            result.append("-" + tokens[i+1])
            i += 2
        else:
            result.append(tokens[i])
            i += 1

    return result


# ================= INFIX → POSTFIX =================
def infix_to_postfix(tokens):
    stack = StackLL()
    postfix = []

    print("\n=== PROSES INFIX → POSTFIX ===")

    for t in tokens:
        print("\nToken:", t)

        if re.match(r'-?\d+\.?\d*', t):
            postfix.append(t)
            print("Masuk ke postfix")

        elif t == "(":
            stack.push(t)

        elif t == ")":
            while stack.peek() != "(":
                postfix.append(stack.pop())
            stack.pop()

        elif is_operator(t):
            while (not stack.is_empty() and
                   precedence(stack.peek()) >= precedence(t)):
                postfix.append(stack.pop())
            stack.push(t)

        print("Stack:", end=" ")
        stack.print_stack()
        print("Postfix:", postfix)

    while not stack.is_empty():
        postfix.append(stack.pop())

    print("\nFinal Postfix:", postfix)
    return postfix


# ================= EVALUASI =================
def evaluate_postfix(postfix):
    stack = StackLL()

    print("\n=== PROSES EVALUASI POSTFIX ===")

    for t in postfix:
        if re.match(r'-?\d+\.?\d*', t):
            stack.push(float(t))
            print("Push", t)

        else:
            try:
                b = stack.pop()
                a = stack.pop()
            except:
                raise Exception("Error: operand kurang!")

            if t == "+": result = a + b
            elif t == "-": result = a - b
            elif t == "*": result = a * b
            elif t == "/":
                if b == 0:
                    raise Exception("Error: pembagian dengan nol!")
                result = a / b
            elif t == "^": result = a ** b

            print(f"Hitung: {a} {t} {b} = {result}")
            stack.push(result)

        print("Stack:", end=" ")
        stack.print_stack()

    return stack.pop()


# ================= MAIN =================
while True:
    print("\n=================================")
    expr = input("Masukkan ekspresi aritmatika : ")

    try:
        tokens = tokenize(expr)
        postfix = infix_to_postfix(tokens)
        result = evaluate_postfix(postfix)

        print("\nHASIL AKHIR =", result)

    except Exception as e:
        print("ERROR:", e)

    ulang = input("\nHitung lagi? (y/n): ").lower()
    if ulang != "y":
        break

print("Program selesai")

File: StrukturDataStack/StackExpressionLinkedList.java
import java.util.*;

class Node {
    String data;
    Node next;

    Node(String data) {
        this.data = data;
        this.next = null;
    }
}

class StackLL {
    Node top;

    boolean isEmpty() {
        return top == null;
    }

    void push(String data) {
        Node newNode = new Node(data);
        newNode.next = top;
        top = newNode;
    }

    String pop() {
        if (isEmpty()) return null;
        String value = top.data;
        top = top.next;
        return value;
    }

    String peek() {
        if (isEmpty()) return null;
        return top.data;
    }

    void printStack() {
        Node temp = top;
        System.out.print("[ ");
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.print("]");
    }
}

public class StackExpressionLinkedList {

    static int precedence(String op) {
        if (op.equals("+") || op.equals("-")) return 1;
        if (op.equals("*") || op.equals("/")) return 2;
        return 0;
    }

    static boolean isOperator(String s) {
        return s.equals("+") || s.equals("-") || s.equals("*") || s.equals("/");
    }

    static String infixToPostfix(String[] tokens) {

        StackLL stack = new StackLL();
        String postfix = "";

        System.out.println("\n=== PROSES INFIX → POSTFIX ===");

        for (String t : tokens) {

            System.out.println("\nToken : " + t);

            if (t.matches("\\d+")) {
                postfix += t;
                System.out.println("Masuk ke postfix");
            }

            else if (t.equals("(")) {
                stack.push(t);
                System.out.println("Push '(' ke stack");
            }

            else if (t.equals(")")) {

                while (!stack.peek().equals("(")) {
                    postfix += stack.pop();
                }
                stack.pop();
                System.out.println("Pop sampai '('");
            }

            else if (isOperator(t)) {

                while (!stack.isEmpty() &&
                        precedence(stack.peek()) >= precedence(t)) {

                    postfix += stack.pop();
                }

                stack.push(t);
                System.out.println("Push operator ke stack");
            }

            System.out.print("Stack : ");
            stack.printStack();

            System.out.println("\nPostfix : " + postfix);
        }

        while (!stack.isEmpty()) {
            postfix += stack.pop();
        }

        System.out.println("\nFinal Postfix : " + postfix);

        return postfix;
    }

    static double evaluatePostfix(String postfix) {

        StackLL stack = new StackLL();

        System.out.println("\n=== PROSES EVALUASI POSTFIX ===");

        for (int i = 0; i < postfix.length(); i++) {

            String t = String.valueOf(postfix.charAt(i));

            if (t.matches("\\d")) {

                stack.push(t);
                System.out.println("Push " + t);

                System.out.print("Stack : ");
                stack.printStack();
                System.out.println();
            }

            else {

                double b = Double.parseDouble(stack.pop());
                double a = Double.parseDouble(stack.pop());

                double result = 0;

                switch (t) {
                    case "+": result = a + b; break;
                    case "-": result = a - b; break;
                    case "*": result = a * b; break;
                    case "/": result = a / b; break;
                }

                System.out.println("Pop " + a + " dan " + b);
                System.out.println("Hitung : " + a + " " + t + " " + b + " = " + result);

                stack.push(String.valueOf(result));

                System.out.print("Stack : ");
                stack.printStack();
                System.out.println();
            }
        }

        return Double.parseDouble(stack.pop());
    }

    public static void main(String[] args) {

        Scanner sc = new Scanner(System.in);
        String loop;

        do {

            System.out.println("\n=================================");
            System.out.print("Masukkan ekspresi aritmatika : ");

            String input = sc.nextLine();

            String[] tokens = input.replaceAll("([+\\-*/()])", " $1 ").trim().split("\\s+");

            String postfix = infixToPostfix(tokens);

            double result = evaluatePostfix(postfix);

            System.out.println("\nHASIL AKHIR = " + result);

            System.out.print("\nHitung lagi? (y/n) : ");
            loop = sc.nextLine().toLowerCase();

        } while (loop.equals("y"));

        System.out.println("Program selesai");
    }
}